File: src/server/provider.ts
import type {
  ContentChannel,
  ProviderConnectionResult,
  ProviderKind,
} from "@/lib/app-types";

export type ProviderRuntimeSettings = {
  kind: ProviderKind;
  baseUrl: string;
  model: string;
  apiKey: string;
};

type GenerateInput = {
  topic: string;
  channel: ContentChannel;
  tone: string;
  objective: string;
  businessName: string;
  businessDescription: string;
};

export type GeneratedContent = {
  title: string;
  body: string;
  hashtags: string[];
  rationale: string;
};

function validateBaseUrl(value: string) {
  const url = new URL(value);
  if (!['http:', 'https:'].includes(url.protocol)) {
    throw new Error("Provider URL must use http or https.");
  }
  if (url.username || url.password) {
    throw new Error("Provider URL credentials are not allowed. Use the API key field instead.");
  }
  return url.toString().replace(/\/$/, "");
}

function openAiEndpoint(baseUrl: string, resource: "models" | "chat/completions") {
  const normalized = validateBaseUrl(baseUrl);
  return normalized.endsWith("/v1") ? `${normalized}/${resource}` : `${normalized}/v1/${resource}`;
}

function authorizationHeaders(apiKey: string): Record<string, string> {
  return apiKey ? { Authorization: `Bearer ${apiKey}` } : {};
}

async function fetchJson(url: string, init?: RequestInit, timeoutMs = 25_000) {
  const response = await fetch(url, {
    ...init,
    redirect: "error",
    signal: AbortSignal.timeout(timeoutMs),
    headers: {
      Accept: "application/json",
      ...(init?.headers || {}),
    },
  });
  const payload = (await response.json().catch(() => null)) as Record<string, unknown> | null;
  if (!response.ok) {
    const message =
      (payload?.error as { message?: string } | undefined)?.message ||
      (typeof payload?.message === "string" ? payload.message : "") ||
      `${response.status} ${response.statusText}`;
    throw new Error(message);
  }
  return payload;
}

export async function testProvider(settings: ProviderRuntimeSettings): Promise<ProviderConnectionResult> {
  const startedAt = Date.now();
  try {
    if (settings.kind === "ollama") {
      const payload = await fetchJson(`${validateBaseUrl(settings.baseUrl)}/api/tags`);
      const models = Array.isArray(payload?.models)
        ? payload.models
            .map((item) => (item && typeof item === "object" && "name" in item ? String(item.name) : ""))
            .filter(Boolean)
        : [];
      return {
        ok: true,
        message: models.length ? `Ollama connected. ${models.length} local model(s) found.` : "Ollama connected. Pull a model to generate content.",
        models,
        latencyMs: Date.now() - startedAt,
      };
    }

    const payload = await fetchJson(openAiEndpoint(settings.baseUrl, "models"), {
      headers: authorizationHeaders(settings.apiKey),
    });
    const models = Array.isArray(payload?.data)
      ? payload.data
          .map((item) => (item && typeof item === "object" && "id" in item ? String(item.id) : ""))
          .filter(Boolean)
          .slice(0, 50)
      : [];
    return {
      ok: true,
      message: `Provider connected${models.length ? ` with ${models.length} visible model(s)` : ""}.`,
      models,
      latencyMs: Date.now() - startedAt,
    };
  } catch (error) {
    return {
      ok: false,
      message: error instanceof Error ? error.message : "Provider connection failed.",
      latencyMs: Date.now() - startedAt,
    };
  }
}

function generationPrompt(input: GenerateInput) {
  return [
    "You are the senior social media copywriter inside a human-approved marketing workflow.",
    "Return only valid JSON with this exact shape:",
    '{"title":"short internal title","body":"publish-ready post","hashtags":["#tag"],"rationale":"one sentence explaining the angle"}',
    "Do not invent statistics, testimonials, customers, awards, prices, or guarantees.",
    "Avoid generic AI phrases, excessive punctuation, and engagement bait.",
    `Business: ${input.businessName || "Not provided"}`,
    `Business context: ${input.businessDescription || "Not provided"}`,
    `Channel: ${input.channel}`,
    `Topic: ${input.topic}`,
    `Objective: ${input.objective || "Build useful awareness"}`,
    `Tone: ${input.tone || "Clear and confident"}`,
    "Adapt length, structure, and hashtag count to the selected channel.",
  ].join("\n");
}

function parseGeneratedContent(value: string): GeneratedContent {
  const cleaned = value.trim().replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  const start = cleaned.indexOf("{");
  const end = cleaned.lastIndexOf("}");
  if (start < 0 || end <= start) throw new Error("Model did not return valid JSON content.");
  const parsed = JSON.parse(cleaned.slice(start, end + 1)) as Partial<GeneratedContent>;
  if (typeof parsed.title !== "string" || !parsed.title.trim()) throw new Error("Model response has an invalid title.");
  if (typeof parsed.body !== "string" || !parsed.body.trim()) throw new Error("Model response has an invalid body.");
  if (parsed.hashtags !== undefined && (!Array.isArray(parsed.hashtags) || parsed.hashtags.some((tag) => typeof tag !== "string"))) {
    throw new Error("Model response has invalid hashtags.");
  }
  if (parsed.rationale !== undefined && typeof parsed.rationale !== "string") {
    throw new Error("Model response has an invalid rationale.");
  }
  return {
    title: parsed.title.trim().slice(0, 160),
    body: parsed.body.trim().slice(0, 12_000),
    hashtags: Array.isArray(parsed.hashtags)
      ? parsed.hashtags.map((tag) => tag.trim()).filter(Boolean).slice(0, 20)
      : [],
    rationale: parsed.rationale ? parsed.rationale.trim().slice(0, 500) : "",
  };
}

export async function generateContent(
  settings: ProviderRuntimeSettings,
  input: GenerateInput,
): Promise<GeneratedContent> {
  if (!settings.model) throw new Error("Select a model before generating content.");
  const prompt = generationPrompt(input);

  if (settings.kind === "ollama") {
    const payload = await fetchJson(
      `${validateBaseUrl(settings.baseUrl)}/api/chat`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          model: settings.model,
          stream: false,
          format: "json",
          messages: [{ role: "user", content: prompt }],
          options: { temperature: 0.7 },
        }),
      },
      120_000,
    );
    const message = payload?.message as { content?: string } | undefined;
    return parseGeneratedContent(message?.content || "");
  }

  const payload = await fetchJson(
    openAiEndpoint(settings.baseUrl, "chat/completions"),
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authorizationHeaders(settings.apiKey),
      },
      body: JSON.stringify({
        model: settings.model,
        temperature: 0.7,
        messages: [
          { role: "system", content: "You create factual, brand-safe marketing drafts for human review." },
          { role: "user", content: prompt },
        ],
      }),
    },
    120_000,
  );
  const choices = payload?.choices as Array<{ message?: { content?: string } }> | undefined;
  return parseGeneratedContent(choices?.[0]?.message?.content || "");
}

File: backend/app/services/provider.py
from __future__ import annotations

import json
import time
from typing import Any
from urllib.parse import urlsplit, urlunsplit

import httpx

from app.errors import ExternalServiceError
from app.schemas import GeneratedContent, ProviderConnectionResult


def validate_base_url(value: str) -> str:
    parsed = urlsplit(value.strip())
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise ExternalServiceError("Provider URL must be a valid http or https address.")
    if parsed.username or parsed.password:
        raise ExternalServiceError("Provider URL credentials are not allowed. Use the API key field instead.")
    return urlunsplit((parsed.scheme, parsed.netloc, parsed.path.rstrip("/"), "", ""))


def _openai_endpoint(base_url: str, resource: str) -> str:
    normalized = validate_base_url(base_url)
    return f"{normalized}/{resource}" if normalized.endswith("/v1") else f"{normalized}/v1/{resource}"


async def _request_json(
    url: str,
    *,
    method: str = "GET",
    headers: dict[str, str] | None = None,
    json_body: dict[str, Any] | None = None,
    timeout: float = 25,
) -> dict[str, Any]:
    try:
        async with httpx.AsyncClient(follow_redirects=False, timeout=timeout) as client:
            response = await client.request(method, url, headers={"Accept": "application/json", **(headers or {})}, json=json_body)
    except httpx.HTTPError as error:
        raise ExternalServiceError(f"Provider connection failed ({type(error).__name__}).") from error
    try:
        payload = response.json()
    except ValueError as error:
        raise ExternalServiceError(f"Provider returned a non-JSON response ({response.status_code}).") from error
    if not response.is_success:
        message = ""
        if isinstance(payload, dict):
            nested = payload.get("error")
            if isinstance(nested, dict):
                message = str(nested.get("message") or "")
            message = message or str(payload.get("message") or "")
        raise ExternalServiceError(message or f"Provider returned HTTP {response.status_code}.")
    if not isinstance(payload, dict):
        raise ExternalServiceError("Provider returned an invalid JSON object.")
    return payload


async def test_provider(settings: dict[str, str]) -> ProviderConnectionResult:
    started = time.monotonic()
    try:
        if settings["kind"] == "ollama":
            payload = await _request_json(f"{validate_base_url(settings['base_url'])}/api/tags")
            raw_models = payload.get("models") if isinstance(payload.get("models"), list) else []
            models = [str(item.get("name")) for item in raw_models if isinstance(item, dict) and item.get("name")]
            message = (
                f"Ollama connected. {len(models)} local model(s) found."
                if models
                else "Ollama connected. Pull a model to generate content."
            )
        else:
            headers = {"Authorization": f"Bearer {settings['api_key']}"} if settings["api_key"] else {}
            payload = await _request_json(_openai_endpoint(settings["base_url"], "models"), headers=headers)
            raw_models = payload.get("data") if isinstance(payload.get("data"), list) else []
            models = [str(item.get("id")) for item in raw_models if isinstance(item, dict) and item.get("id")][:50]
            message = f"Provider connected{' with ' + str(len(models)) + ' visible model(s)' if models else ''}."
        return ProviderConnectionResult(
            ok=True,
            message=message,
            models=models,
            latency_ms=round((time.monotonic() - started) * 1_000),
        )
    except ExternalServiceError as error:
        return ProviderConnectionResult(
            ok=False,
            message=error.message,
            latency_ms=round((time.monotonic() - started) * 1_000),
        )


def _generation_prompt(request: dict[str, Any], workspace: dict[str, str]) -> str:
    return "\n".join(
        [
            "You are the senior social media copywriter inside a human-approved marketing workflow.",
            "Return only valid JSON with this exact shape:",
            '{"title":"short internal title","body":"publish-ready post","hashtags":["#tag"],"rationale":"one sentence explaining the angle"}',
            "Do not invent statistics, testimonials, customers, awards, prices, or guarantees.",
            "Avoid generic AI phrases, excessive punctuation, and engagement bait.",
            f"Business: {workspace['business_name'] or 'Not provided'}",
            f"Business context: {workspace['business_description'] or 'Not provided'}",
            f"Channel: {request['channel']}",
            f"Topic: {request['topic']}",
            f"Objective: {request['objective'] or 'Build useful awareness'}",
            f"Tone: {request['tone'] or 'Clear and confident'}",
            "Adapt length, structure, and hashtag count to the selected channel.",
        ]
    )


def _parse_content(value: str) -> GeneratedContent:
    cleaned = value.strip()
    if cleaned.startswith("```"):
        cleaned = cleaned.split("\n", 1)[1] if "\n" in cleaned else cleaned[3:]
        cleaned = cleaned.rsplit("```", 1)[0]
    start, end = cleaned.find("{"), cleaned.rfind("}")
    if start < 0 or end <= start:
        raise ExternalServiceError("Model did not return valid JSON content.")
    try:
        payload = json.loads(cleaned[start : end + 1])
    except json.JSONDecodeError as error:
        raise ExternalServiceError("Model did not return valid JSON content.") from error
    if not isinstance(payload, dict):
        raise ExternalServiceError("Model returned an invalid content object.")
    title = str(payload.get("title") or "").strip()[:160]
    body = str(payload.get("body") or "").strip()[:12_000]
    raw_tags = payload.get("hashtags") if isinstance(payload.get("hashtags"), list) else []
    hashtags = [str(tag).strip()[:80] for tag in raw_tags if str(tag).strip()][:20]
    rationale = str(payload.get("rationale") or "").strip()[:500]
    if not title or not body:
        raise ExternalServiceError("Model response is missing a title or body.")
    return GeneratedContent(title=title, body=body, hashtags=hashtags, rationale=rationale)


async def generate_content(
    settings: dict[str, str], request: dict[str, Any], workspace: dict[str, str]
) -> GeneratedContent:
    if not settings["model"]:
        raise ExternalServiceError("Select a model before generating content.")
    prompt = _generation_prompt(request, workspace)
    if settings["kind"] == "ollama":
        payload = await _request_json(
            f"{validate_base_url(settings['base_url'])}/api/chat",
            method="POST",
            headers={"Content-Type": "application/json"},
            json_body={
                "model": settings["model"],
                "stream": False,
                "format": "json",
                "messages": [{"role": "user", "content": prompt}],
                "options": {"temperature": 0.7},
            },
            timeout=120,
        )
        message = payload.get("message") if isinstance(payload.get("message"), dict) else {}
        return _parse_content(str(message.get("content") or ""))

    headers = {"Content-Type": "application/json"}
    if settings["api_key"]:
        headers["Authorization"] = f"Bearer {settings['api_key']}"
    payload = await _request_json(
        _openai_endpoint(settings["base_url"], "chat/completions"),
        method="POST",
        headers=headers,
        json_body={
            "model": settings["model"],
            "temperature": 0.7,
            "messages": [
                {"role": "system", "content": "You create factual, brand-safe marketing drafts for human review."},
                {"role": "user", "content": prompt},
            ],
        },
        timeout=120,
    )
    choices = payload.get("choices") if isinstance(payload.get("choices"), list) else []
    message = choices[0].get("message") if choices and isinstance(choices[0], dict) else {}
    return _parse_content(str(message.get("content") or "") if isinstance(message, dict) else "")
